### PR TITLE
[TG-1404] Specific implementation of cover basic block for Java

### PR DESCRIPTION
**Pull request by romainbrenguier**
The current implementation of cover_basic_blockt does a first pass on the goto-program without using the bytecode index information, and then later select instruction for bytecode indices.
The process is a bit complicated and sometimes instruments instruction that are after an IF of the same bytecode indices, leading to wrong information about bytecode coverage.
This PR intruduce an alternative cover_basice_block_javat, which for Java use the bytecode information to determine what are the blocks of the program, and selects the first instruction with that bytecode index as a representative.
This strategy is simple and ensures each goal corresponds to a bytecode index coverage.

### CODE REVIEW

File: src/goto-instrument/cover_basic_blocks.cpp
Comment (tautschnig) [inline src/goto-instrument/cover_basic_blocks.cpp:90]: Why all these changes to a `std::size_t` when the map itself (`block_mapt`) still maps to `unsigned`? And the commit message hasn't got a single word about this?!

File: src/goto-instrument/cover_instrument_location.cpp
Comment (tautschnig) [inline src/goto-instrument/cover_instrument_location.cpp:25]: Wrong commit?

File: src/goto-instrument/cover_basic_blocks.cpp
Comment (tautschnig) [inline src/goto-instrument/cover_basic_blocks.cpp]: Using `at` together with the above `INVARIANT` is unnecessary. Use `block_infos[block_nr]` or drop the `INVARIANT`. The same is true below.

File: src/goto-instrument/cover_basic_blocks.h
Comment (tautschnig) [inline src/goto-instrument/cover_basic_blocks.h]: Use `irep_idt` (and `irep_id_hash`), never use `dstringt` directly.

File: src/goto-instrument/cover_basic_blocks.cpp
Comment (tautschnig) [inline src/goto-instrument/cover_basic_blocks.cpp]: That's two lookups in quick succession. Instead:
````c++
auto &entry = index_to_block.insert({bytecode_index, block_infos.size() - 1});
if(entry.second)
  block_infos.push_back(it);
````

File: src/goto-instrument/cover.cpp
Comment (peterschrammel) [inline src/goto-instrument/cover.cpp]: const &

File: src/goto-instrument/cover_instrument.h
Comment (peterschrammel) [inline src/goto-instrument/cover_instrument.h]: Changing the forward declaration below to cover_blocks_baset should be sufficient.

File: src/goto-instrument/cover_basic_blocks.cpp
Comment (peterschrammel) [inline src/goto-instrument/cover_basic_blocks.cpp]: Where are the covered_lines set?

**Review by peterschrammel — CHANGES_REQUESTED** [review]

File: src/goto-instrument/cover_basic_blocks.cpp
Comment (romainbrenguier) [reply src/goto-instrument/cover_basic_blocks.cpp:90]: This was to fit the function type of the interface. I guess I will add a new commit at the beginning to do these type changes.

File: src/goto-instrument/cover_basic_blocks.cpp
Comment (romainbrenguier) [reply src/goto-instrument/cover_basic_blocks.cpp]: Well spotted (except size-1 needs to be changed to size)

File: src/goto-instrument/cover_basic_blocks.cpp
Comment (romainbrenguier) [reply src/goto-instrument/cover_basic_blocks.cpp]: They are not. Because I don't see how instructions with same bytecode index can correspond to different lines?

File: src/goto-instrument/cover_instrument_location.cpp
Comment (romainbrenguier) [reply src/goto-instrument/cover_instrument_location.cpp:25]: I added an initial commit for these changes

File: src/goto-instrument/cover_basic_blocks.cpp
Comment (peterschrammel) [reply src/goto-instrument/cover_basic_blocks.cpp]: They must be set nonetheless in order not to change the output format.

File: src/goto-instrument/cover_basic_blocks.h
Comment (romainbrenguier) [reply src/goto-instrument/cover_basic_blocks.h]: @tautschnig Alright if you say so. But may I ask, irep_idt is just a typedef dstringt, is there a reason we keep the two separated?

Comment (tautschnig) [reply src/goto-instrument/cover_basic_blocks.h]: The macro `USE_DSTRING` will make an `irep_idt` be a `dstringt`, otherwise an `irep_idt` is a `std::string`. It's a different question whether disabling `USE_DSTRING` is actually beneficial.

**Review by LAJW — APPROVED** [review]

File: src/goto-instrument/cover_basic_blocks.cpp
Comment (romainbrenguier) [reply src/goto-instrument/cover_basic_blocks.cpp]: Oh right, I thought this was the same as get_line. This should be fixed in the last fixup commit https://github.com/diffblue/cbmc/pull/1830/commits/48fc65ceb2d5ec425e074e37a809237f4fa85a13

File: src/goto-instrument/cover_basic_blocks.h
Comment (romainbrenguier) [reply src/goto-instrument/cover_basic_blocks.h]: I just try commenting out the `#define USE_DSTRING` and cbmc does not compile. So I guess it has not been tested for some time.

Comment (tautschnig) [reply src/goto-instrument/cover_basic_blocks.h]: Would you mind creating an issue, possibly including your build log? Certainly it hasn't been tested in a long time, and it should be properly evaluated whether it makes sense at all anymore. But there doesn't really seem any urgency to do this.

File: src/goto-instrument/cover_basic_blocks.h
Comment (thk123) [inline src/goto-instrument/cover_basic_blocks.h:21]: Nit: move this into its own file

File: src/goto-instrument/cover.cpp
Comment (thk123) [inline src/goto-instrument/cover.cpp]: The valid values for `mode` are `ID_java` or anything else? Consider an enum if you don't want to use a `bool enable_java_blocks`

File: src/goto-instrument/cover.cpp
Comment (thk123) [inline src/goto-instrument/cover.cpp:332]: Eek local variable called `config` that's going to bite someone one day...

File: src/goto-instrument/cover_basic_blocks.h
Comment (thk123) [reply src/goto-instrument/cover_basic_blocks.h]: I've always hoped this would be fixed as IDEs have much better support for std::string - I thought I had raised an issue but that seems to be all in my head

File: src/goto-instrument/cover_basic_blocks.cpp
Comment (thk123) [inline src/goto-instrument/cover_basic_blocks.cpp]: Nit: probably a PRECONDITION

File: src/goto-instrument/cover_basic_blocks.h
Comment (romainbrenguier) [reply src/goto-instrument/cover_basic_blocks.h]: issue created here: https://github.com/diffblue/cbmc/issues/1837

**Review by thk123 — COMMENTED** [review]
It would be good to have some tests (in fact, for free this fixes some problems so tests introduced by #1812 plus something along the lines of the tests introduced in diffblue/test-gen#1474

File: src/goto-instrument/cover.cpp
Comment (peterschrammel) [reply src/goto-instrument/cover.cpp]: `mode` is a standard concept used in the symbol table.

**Comment by romainbrenguier** [comment]
I modified a test in https://github.com/diffblue/cbmc/pull/1830/commits/1083f74776d6d9039078ef656f219e988a876b62
which should show the difference between before and after these changes.
In https://github.com/diffblue/cbmc/pull/1812 I only see one unit test, and what is tested seems complicated compared to what the new instrumentation is doing. 
It difficult to have test showing the previous implementation would fail, because it depends on some non-predictable behaviour (like the order of some instructions which are added during instrumentation).
I will activate test similar to https://github.com/diffblue/test-gen/pull/1474 with the corresponding submodule update.

File: src/goto-instrument/cover_basic_blocks.h
Comment (romainbrenguier) [reply src/goto-instrument/cover_basic_blocks.h:21]: I don't know, this is just an interface so very little code is added and this file is relatively small...

**Review by thk123 — APPROVED** [review]
@mgudemann do you agree that the unit test in #1812 is not relevant for this PR? 

@romainbrenguier In any case, it seemed like the virtual function resolution was consistently wrong (block would always go inside the block of the first if) - can this perhaps be used to have a regression/unit test?

**Comment by mgudemann** [comment]
@thk123 I think I would make sense to adapt the test which would mainly mean removing the test on the added marker. But testing that there is no `ASSERT` after the initial guarded GOTO would still make sense, wouldn't it?
https://github.com/diffblue/cbmc/pull/1812/files#diff-0742ad545166f7e1ac81c0e158edb2b1R158

**Comment by romainbrenguier** [comment]
@thk123 @mgudemann yes this makes sense, I will try adding the unit test

**Comment by romainbrenguier** [comment]
@thk123 @mgudemann I added the unit test

File: unit/java_bytecode/java_virtual_functions/virtual_functions.cpp
Comment (thk123) [inline unit/java_bytecode/java_virtual_functions/virtual_functions.cpp]: My feeling is you should do `std::next(goto_program.instructions.begin())` but then the require would be that the program includes at least two source locations, but currently you check the first one is not the end, then advance one and then dereference it, which might be a problem as it might be at the end. Using `std::next` when init'ing would prevent this.

File: unit/java_bytecode/java_virtual_functions/virtual_functions.cpp
Comment (thk123) [inline unit/java_bytecode/java_virtual_functions/virtual_functions.cpp]: Prefer `REQUIRE_FALSE` when checking a negative

File: unit/java_bytecode/java_virtual_functions/virtual_functions.cpp
Comment (thk123) [inline unit/java_bytecode/java_virtual_functions/virtual_functions.cpp]: Prefer `REQUIRE_FALSE` when checking a negative

**Review by thk123 — APPROVED** [review]
Unit test seems good - couple of suggestions but otherwise LGTM

**Review by peterschrammel — APPROVED** [review]
I think that the internal implementation could be further simplified and unified, but we can do that in a later PR when integrating https://github.com/diffblue/cbmc/pull/1439. Let's go ahead with that to solve first things first.

File: src/goto-instrument/cover.cpp
Comment (chrisr-diffblue) [inline src/goto-instrument/cover.cpp]: Doxygen needs adding for the extra parameter, and as a minor nit, I might be tempted to put the new `mode` parameter before the `message_handler` parameter as the diff would be a fraction smaller, and it groups the parameters that have more meaning together i.e. leaving just the UI parameter tacked on the end.

**Review by chrisr-diffblue — APPROVED** [review]
One minor comment, but generally looks good so I'll approve and let you consider whether you agree with my comment or not.

**Comment by romainbrenguier** [comment]
I have not merge this PR yet because I have seen a test failing on another repository after that change.  This other PR fixes the problem https://github.com/diffblue/cbmc/pull/1866 , so it would be better to wait for this second PR to be merged before merging this one.

**Pull request merged**
